packages/babel-generator/src/node/parentheses.js

```diff
@@ -127,6 +127,17 @@ export function TSTypeAssertion() {
   return true;
 }
 
+export function TSUnionType(node: Object, parent: Object): boolean {
+  return (
+    t.isTSArrayType(parent) ||
+    t.isTSNullKeyword(parent) ||
```

Comment:
Why? `null` doesn't have any child.

Reply:
I was meaning `isTSOptionalType`, basically I copied the logic from `UnionTypeAnnotation() {}` above, is there a way to tell `isArrayTypeAnnotation` and `isTSArrayType`, `isIntersectionTypeAnnotation` and `isTSIntersectionType`  is similar, so don't have to rewrite the same logic again

Reply:
> is there a way to tell `isArrayTypeAnnotation` and `isTSArrayType`, `isIntersectionTypeAnnotation` and `isTSIntersectionType` is similar, so don't have to rewrite the same logic again

I don't think so: typescript and flow can diverge at any time and we don't have a way to compare their ASTs.

PS. You should also add `TSRestType`.

Reply:
OK, I've added `TSRestType` too, as well as test